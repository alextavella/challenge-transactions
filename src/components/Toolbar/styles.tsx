import { Typography } from '@material-ui/core';
import styled from 'styled-components';

export const Title = styled(Typography)`
  width: 100%;
  margin-right: 48px;
`;
