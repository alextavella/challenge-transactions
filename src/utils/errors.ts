import { ValidationError } from 'yup';

interface Error {
  [key: string]: string;
}

export function getValidationErrors(errors: ValidationError): any {
  return errors.inner.reduce((acc: Error, error: ValidationError) => {
    return { ...acc, [error.path]: error.message };
  }, {});
}
