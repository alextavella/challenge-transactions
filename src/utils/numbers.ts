export const formatCurrency = (value: number): string => {
  const formatted = new Intl.NumberFormat([], {
    style: 'currency',
    currency: 'BRL',
    minimumFractionDigits: 2,
  }).format(value);

  return formatted
    .replace('.', '#')
    .replace(/,/g, '.')
    .replace('#', ',')
    .replace('R$', 'R$ ');
};

export const unformatCurrency = (value: string): number => {
  return Number(
    value
      .replace('R$ ', '')
      .replace(',', '#')
      .replace(/\./g, '')
      .replace('#', '.'),
  );
};
