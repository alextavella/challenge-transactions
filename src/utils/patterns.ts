import { ValidationRules } from 'react-hook-form';

export interface MaskRule {
  rules: ValidationRules;
  transform: (text: string) => string;
}

export const CPF: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/^([0-9]{3}).([0-9]{3}).([0-9]{3})-([0-9]{2})$/),
      message: 'CPF inválido',
    },
  },
  transform: (text: string = ''): string => {
    const result = text
      .replace(/\D/g, '')
      .match(/(\d{0,3})(\d{0,3})(\d{0,3})(\d{0,2})/)!;

    const part1 = result[2] ? `${result[1]}.${result[2]}` : result[1];
    const part2 = result[3] ? `.${result[3]}` : '';
    const part3 = result[4] ? `-${result[4]}` : '';

    return `${part1}${part2}${part3}`;
  },
};

export const CREDIT_CARD_NUMBER: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/^([0-9]{4})\s([0-9]{4})\s([0-9]{4})\s([0-9]{4})$/),
      message: 'Número do cartão inválido',
    },
  },
  transform: (text: string = ''): string => {
    const result = text
      .replace(/\D/g, '')
      .match(/(\d{0,4})(\d{0,4})(\d{0,4})(\d{0,4})/)!;

    const part1 = result[2] ? `${result[1]} ${result[2]}` : result[1];
    const part2 = result[3] ? ` ${result[3]}` : '';
    const part3 = result[4] ? ` ${result[4]}` : '';

    return `${part1}${part2}${part3}`;
  },
};

export const CREDIT_CARD_EXPIRATION_DATE: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/^(0[1-9]|1[012])\/?(20[2-9][0-9])$/),
      message: 'Data de expiração inválida',
    },
  },
  transform: (text: string = ''): string => {
    const result = text.replace(/\D/g, '').match(/(\d{0,2})(\d{0,4})/)!;

    return result[2] ? `${result[1]}/${result[2]}` : result[1];
  },
};

export const CREDIT_CARD_CVV: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/[0-9]{3}/),
      message: 'CVV inválido',
    },
  },
  transform: (text: string = ''): string => {
    return text.substr(0, 3);
  },
};

export const CURRENCY: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/([0-9]\.?)?[0-9]+,[0-9]{2}/),
      message: 'Valor inválido',
    },
  },
  transform: (text: string = ''): string => {
    let number = text.replace(/[^0-9]/g, '');

    if (number.length === 0) number = '0.00';
    else if (number.length === 1) number = '0.0' + number;
    else if (number.length === 2) number = '0.' + number;
    else
      number =
        number.substring(0, number.length - 2) +
        '.' +
        number.substring(number.length - 2, number.length);

    number = Number(number).toFixed(2);
    number = number.replace(/\./g, ',');

    const x = number.split(',');
    const x1 = x[0];
    const x2 = x.length > 1 ? ',' + x[1] : '';

    const formatAmountNoDecimals = (number: string) => {
      var rgx = /(\d+)(\d{3})/;
      while (rgx.test(number)) {
        number = number.replace(rgx, `$1.$2`);
      }
      return number;
    };

    return `R$ ` + formatAmountNoDecimals(x1) + x2;
  },
};
