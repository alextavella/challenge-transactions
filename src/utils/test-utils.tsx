import { render } from '@testing-library/react';
import React from 'react';
import MainProvider from '../providers';

const AllTheProviders: React.FC = ({ children }) => {
  return <MainProvider>{children}</MainProvider>;
};

const customRender = (ui: React.ReactElement, options?: any) =>
  render(ui, {
    wrapper: AllTheProviders,
    ...options,
  });

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
