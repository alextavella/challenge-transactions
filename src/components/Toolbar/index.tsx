import { IconButton, Toolbar } from '@material-ui/core';
import ArrowBackIcon from '@material-ui/icons/ArrowBack';
import React from 'react';
import { Title } from './styles';

interface MyToolbarProps {
  title: string;
  onBack: () => void;
}

const MyToolbar: React.FC<MyToolbarProps> = ({ title, onBack }) => {
  return (
    <Toolbar variant="dense">
      <IconButton
        data-testid="back-button"
        edge="start"
        color="primary"
        onClick={() => onBack()}
      >
        <ArrowBackIcon />
      </IconButton>
      <Title variant="h6" color="inherit">
        {title}
      </Title>
    </Toolbar>
  );
};

export default MyToolbar;
