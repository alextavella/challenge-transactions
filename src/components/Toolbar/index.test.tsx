import React from 'react';
import Toolbar from '.';
import { fireEvent, render, screen, cleanup } from '../../utils/test-utils';

describe('<Toolbar />', () => {
  afterEach(() => {
    cleanup();
  });

  it('should be able to render correctly', () => {
    // Arrange
    const { baseElement } = render(<Toolbar title="Title" onBack={() => {}} />);

    // Assert
    expect(baseElement).toMatchSnapshot();
  });

  it('should be able to call callback function when click on back button', () => {
    // Arrange
    const spyBack = jest.fn();

    render(<Toolbar title="Title" onBack={spyBack} />);

    // Act
    fireEvent.click(screen.getByTestId('back-button'));

    // Assert
    expect(spyBack).toHaveBeenCalled();
  });
});
