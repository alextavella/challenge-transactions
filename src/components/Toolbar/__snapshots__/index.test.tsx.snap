// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toolbar /> should be able to render correctly 1`] = `
<body>
  <div>
    <div
      class="MuiToolbar-root MuiToolbar-dense MuiToolbar-gutters"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-edgeStart"
        data-testid="back-button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <h6
        class="MuiTypography-root sc-bdfBwQ iMDlDH MuiTypography-h6 MuiTypography-colorInherit"
      >
        Title
      </h6>
    </div>
  </div>
</body>
`;
